cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go

```diff
@@ -21,38 +21,134 @@ import (
 	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	kube_client "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
 )
 
 // AwsCloudProvider implements CloudProvider interface.
 type AwsCloudProvider struct {
 	awsManager *AwsManager
+	asgMutex   sync.Mutex
 	asgs       []*Asg
 }
 
 // BuildAwsCloudProvider builds CloudProvider implementation for AWS.
-func BuildAwsCloudProvider(awsManager *AwsManager, specs []string) (*AwsCloudProvider, error) {
+func BuildAwsCloudProvider(awsManager *AwsManager, specs []string, kubeClient kube_client.Interface, autoDiscoverNodeGroup bool, scanInterval time.Duration) (*AwsCloudProvider, error) {
 	aws := &AwsCloudProvider{
 		awsManager: awsManager,
 		asgs:       make([]*Asg, 0),
 	}
-	for _, spec := range specs {
-		if err := aws.addNodeGroup(spec); err != nil {
-			return nil, err
+	if autoDiscoverNodeGroup {
+		go func() {
+			for {
+				select {
+				case <-time.After(scanInterval):
+					{
+						selector := "master!=true"
+						listAll := apiv1.ListOptions{LabelSelector: selector}
+						nl, err := kubeClient.Core().Nodes().List(listAll)
+						if err != nil {
+							glog.Errorf("err: %s\n", err)
+						}
+
+						var nodesList []*string
+
+						for _, n := range nl.Items {
+							glog.V(4).Infof("Considering node: %s", n.Name)
+							nodesList = append(nodesList, &n.Name)
+						}
+						AutoDiscoverNodeGroup(aws, nodesList)
+					}
+				}
+			}
+		}()
+	} else {
+		for _, spec := range specs {
+			if err := aws.addNodeGroup(spec); err != nil {
+				return nil, err
+			}
 		}
 	}
 	return aws, nil
 }
 
+//AutoDiscoverNodeGroup will auto update the autoscaler group fetching the information directly from the AWS API.
+func AutoDiscoverNodeGroup(aws *AwsCloudProvider, nodesList []*string) error {
+	params := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws_util.String("private-dns-name"),
+				Values: nodesList,
+			},
+		},
+	}
+	instances, err := aws.awsManager.ec2Service.DescribeInstances(params)
+	if err != nil {
+		glog.Errorf("Error describing EC2 instances: %s\n", err)
+		return err
+	}
+
+	//make a "uniq" on the names so that we don't consider autoscaling groups twice
+	asgs := make(map[string]bool, 0)
+	names := make([]*string, 0)
+	for idx := range instances.Reservations {
+		for _, inst := range instances.Reservations[idx].Instances {
+			for _, tag := range inst.Tags {
+				if *(tag.Key) == "aws:autoscaling:groupName" {
+					if _, ok := asgs[*(tag.Value)]; !ok {
+						names = append(names, tag.Value)
+					}
+					asgs[*(tag.Value)] = true
+				}
+			}
+		}
+	}
+
+	request := &autoscaling.DescribeAutoScalingGroupsInput{
+		AutoScalingGroupNames: names,
+	}
+	asgGroups, err := aws.awsManager.autoscalingService.DescribeAutoScalingGroups(request)
+	if err != nil {
+		glog.Errorf("Error describing autoscaling groups: %s\n", err)
+		return err
+	}
+
+	aws.asgMutex.Lock()
+	defer aws.asgMutex.Unlock()
+	aws.asgs = aws.asgs[:0]
```

Comment:
Ahh I see, that is odd, I would ask if you could possibly fix that but it seems like the other cloud providers do that too. 

GCE: https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/cloudprovider/gce/gce_manager.go#L48 and https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go#L31